DirectXTex/DirectXTexSwizzle.cpp

```diff
@@ -0,0 +1,285 @@
+
+    if (IsTypeless(srcImage.format) || IsPlanar(srcImage.format) || IsPalettized(srcImage.format))
+        return HRESULT_E_NOT_SUPPORTED;
+
+    HRESULT hr = result.Initialize2D(srcImage.format, srcImage.width, srcImage.height, 1, 1);
+    if (FAILED(hr))
+        return hr;
+
+    const uint8_t* sptr = srcImage.pixels;
+    if (!sptr)
+        return E_POINTER;
+
+    uint8_t* dptr = result.GetImages()[0].pixels;
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
+
+    uint32_t xBytesMask = 0b1010101010101010;
+    uint32_t yBytesMask = 0b0101010101010101;
+
+    size_t height = IsCompressed(srcImage.format) ? (srcImage.height + 3) / 4 : srcImage.height;
+    size_t width  = IsCompressed(srcImage.format) ? (srcImage.width  + 3) / 4 : srcImage.width;
+
+    if (toSwizzle)
+    {
+        size_t rowPitch = srcImage.rowPitch;
+        for (size_t y = 0; y < height; y++)
+        {
+            for (size_t x = 0; x < width; x++)
+            {
+                uint32_t swizzleIndex = deposit_bits(x, xBytesMask) + deposit_bits(y, yBytesMask);
+                size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+                size_t rowMajorOffset = y * rowPitch + x * bytesPerPixel;
+
+                const uint8_t* sourcePixelPointer = sptr + rowMajorOffset;
+                uint8_t* destPixelPointer = dptr + swizzleOffset;
+                memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+            }
+        }
+    }
+    else
+    {
+        size_t rowPitch = result.GetImages()[0].rowPitch;
+        for (size_t swizzleIndex = 0; swizzleIndex < (width * height); swizzleIndex++)
+        {
+            size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+            uint32_t destX = extract_bits(swizzleIndex, xBytesMask);
+            uint32_t destY = extract_bits(swizzleIndex, yBytesMask);
+            size_t rowMajorOffset = destY * rowPitch + destX * bytesPerPixel;
+
+            const uint8_t* sourcePixelPointer = sptr + swizzleOffset;
+            uint8_t* destPixelPointer = dptr + rowMajorOffset;
+            memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+        }
+    }
+
+    return S_OK;
+}
+
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle(const Image* srcImages, size_t nimages, const TexMetadata& metadata, bool toSwizzle, ScratchImage& result) noexcept
+{
+    HRESULT hr = result.Initialize2D(srcImages[0].format, srcImages[0].width, srcImages[0].height, nimages, 1);
+
+    if (!srcImages || !nimages || !IsValid(metadata.format) || nimages > metadata.mipLevels || !result.GetImages())
+        return E_INVALIDARG;
+
+    if (metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format))
+        return HRESULT_E_NOT_SUPPORTED;
+
+    if (srcImages[0].format != metadata.format || srcImages[0].width != metadata.width || srcImages[0].height != metadata.height)
```

Comment:
For `ScratchImage` you can guarantee that all the images are the same format since they are created that way by code we control, but for `Image` array input I validate that they are the same as the created result as part of the loop. Our only real concern here is making sure that the `metadata` passed properly describes the input array of images that we were given.

```
    for (size_t imageIndex = 0; imageIndex < nimages; ++imageIndex)
    {
        const Image& src = srcImages[index];
        if (src.format != metadata.format)
        {
            result.Release();
            return E_FAIL;
        }

        if ((src.width > UINT32_MAX) || (src.height > UINT32_MAX))
        {
            result.Release();
            return E_FAIL;
        }
       
        const uint8_t* sptr = src.pixels;
        if (!sptr)
        {
            result.Release();
            return E_POINTER;
        }

        const Image& dst = result.GetImages()[index];
        assert(dst.format == metadata.format);

        if (src.width != dst.width || src.height != dst.height)
        {
            result.Release();
            return E_FAIL;
        }

        uint8_t* dptr = dest.pixels;
        if (!dptr)
        {
            result.Release();
            return E_POINTER;
        }
```